shared/Microsoft.Extensions.Certificates.Generation.Sources/CertificateManager.cs

```diff
@@ -0,0 +1,414 @@
+        }
+
+        public EnsureCertificateResult EnsureAspNetCoreApplicationTokensDevelopmentCertificate(
+            DateTimeOffset notBefore,
+            DateTimeOffset notAfter,
+            string path = null,
+            bool trust = false,
+            bool includePrivateKey = false,
+            string password = null)
+        {
+            return EnsureCertificate(notBefore, notAfter, CertificatePurpose.Signing, path, trust, includePrivateKey, password);
+        }
+
+        public EnsureCertificateResult EnsureCertificate(
+            DateTimeOffset notBefore,
+            DateTimeOffset notAfter,
+            CertificatePurpose purpose,
+            string path = null,
+            bool trust = false,
+            bool includePrivateKey = false,
+            string password = null)
+        {
+            if (purpose == CertificatePurpose.All)
+            {
+                throw new ArgumentException("The certificate must have a specific purpose.");
+            }
+
+            var certificates = ListCertificates(purpose, StoreName.My, StoreLocation.CurrentUser, isValid: true).Concat(
+                ListCertificates(purpose, StoreName.My, StoreLocation.LocalMachine, isValid: true));
+            var result = EnsureCertificateResult.Succeeded;
+
+            X509Certificate2 certificate = null;
+            if (certificates.Count() > 0)
+            {
+                certificate = certificates.FirstOrDefault();
+                result = EnsureCertificateResult.ValidCertificatePresent;
+            }
+            else
+            {
+                try
+                {
+                    switch (purpose)
+                    {
+                        case CertificatePurpose.All:
+                            throw new InvalidOperationException("The certificate must have a specific purpose.");
+                        case CertificatePurpose.HTTPS:
+                            certificate = CreateAspNetCoreHttpsDevelopmentCertificate(notBefore, notAfter);
+                            break;
+                        case CertificatePurpose.Signing:
+                            certificate = CreateApplicationTokenSigningDevelopmentCertificate(notBefore, notAfter);
+                            break;
+                        default:
+                            throw new InvalidOperationException("The certificate must have a purpose.");
+                    }
+                }
+                catch
+                {
+                    return EnsureCertificateResult.ErrorCreatingTheCertificate;
+                }
+
+                try
+                {
+                    certificate = SaveCertificateInStore(certificate, StoreName.My, StoreLocation.CurrentUser);
+                }
+                catch
+                {
+                    return EnsureCertificateResult.ErrorSavingTheCertificateIntoTheCurrentUserPersonalStore;
+                }
+            }
+            if (path != null)
```

Comment:
Why is exporting the key a requirement for being valid?

Reply:
We need to export the key into a PFX for container/unix scenarios, we don't want to take into account any certificate that doesn't cover all our scenarios (has not expired, has a private key and the key is exportable)

The idea is that tooling will run a single command/gesture and ensure there is a valid certificate (not expired, with an exportable private key) that can be used to https and token signing. For containers scenarios the certificate will be exported into a pfx file somewhere on disk and that folder will be mapped so that it can get pick up by kestrel